frontend/kubernetes/plugin.go

```diff
@@ -1508,3 +1542,178 @@ func (p *Plugin) updatePVCSize(
 
 	return pvcUpdated, nil
 }
+
+func (p *Plugin) addVolumeSnapshot(obj interface{}) {
+	volumeSnapshot, ok := obj.(*k8ssnapshotv1alpha1.VolumeSnapshot)
+	if !ok {
+		log.Errorf("Kubernetes frontend expected VolumeSnapshot CR; handler got %v", obj)
+		return
+	}
+	p.processAddedVolumeSnapshot(volumeSnapshot)
+}
+
+func (p *Plugin) processAddedVolumeSnapshot(volSnap *k8ssnapshotv1alpha1.VolumeSnapshot) {
+
+	// Validate the VolumeSnapshot CR
+	if volSnap.Status.Error != nil {
+		log.WithFields(log.Fields{
+			"volumeSnapshot": volSnap.Name,
+		}).Error("Kubernetes frontend found error set on the VolumeSnapshot")
+		return
+	}
+	if volSnap.Spec.Source.Kind != "PersistentVolumeClaim" {
+		log.WithFields(log.Fields{
+			"volumeSnapshot": volSnap.Name,
+		}).Debug("Kubernetes frontend ignored this VolumeSnapshot CR as " +
+			"the source is not a PVC")
+		return
+	}
+
+	log.WithFields(log.Fields{
+		"volumeSnapshot": volSnap.Name,
+		"namespace":      volSnap.Namespace,
+		"sourceKind":     volSnap.Spec.Source.Kind,
+		"sourceName":     volSnap.Spec.Source.Name,
+	}).Debug("Kubernetes frontend got notified of a VolumeSnapshot CR.")
+
+	// Validate the source PVC
+	// 1. PVC and VolumeSnapshotCR must be in the same namespace
+	sourcePVC := volSnap.Spec.Source.Name
+	k8sClient, err := p.getNamespacedKubeClient(&p.kubeConfig, volSnap.Namespace)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"namespace": volSnap.Namespace,
+			"error":     err.Error(),
+		}).Error("Kubernetes frontend couldn't create a client to namespace!")
+		return
+	}
+	claim, err := k8sClient.GetPVC(sourcePVC, metav1.GetOptions{})
+	if err != nil {
+		err = fmt.Errorf("snapshotting a PVC requires the PVC to be in the same namespace as the VolumeSnapshot CR")
+		log.WithFields(log.Fields{
+			"volumeSnapshot": volSnap.Name,
+			"sourcePVC":      sourcePVC,
+		}).Errorf("Kubernetes frontend detected an invalid configuration "+
+			"for creating snapshot from a PVC: %v", err.Error())
+		return
+	}
+
+	// 2. Trident should be set as the provisioner for the PVC
+	provisioner := getClaimProvisioner(claim)
+	if provisioner != AnnOrchestrator &&
+		GetPersistentVolumeClaimClass(claim) != "" {
```

Comment:
Please excuse my ignorance. What does this check do?

Reply:
This is checking if the source PVC was provisioned by Trident.